### DISCUSSION TITLE
Welcome to prana_rc Discussions!

### DISCUSSION THREAD

**Discussion by corvis** [corvis/prana_rc#13] [General]
## 👋 Welcome!
  We’re using Discussions as a place to connect with other members of our community. We hope that you:
  * Ask questions you’re wondering about.
  * Share ideas.
  * Engage with other community members.
  * Welcome others and are open-minded. Remember that this is a community we
  build together 💪.

  To get started, comment below with an introduction of yourself and tell us about what you do with this community.

**Comment by francesco-re-1107** [comment]
Hi,
first of all thank you for your hard work in reversing Prana protocol.

I've made a few changes to your project to be able to read the built-in sensors of my Prana serie 160.
I don't know if you already know how to fetch those information (maybe it depends on the model), however I would like to share with you how I managed to read those values. Here you can find the commit: https://github.com/francesco-re-1107/prana_rc/commit/1c04384181d16c17a25634460b443298d288c740

Maybe it could be useful for someone

**Reply by corvis** [reply]
Hey @francesco-re-1107,
That's awesome. I'll definitely incorporate the changes into the mainline code base. 

May I ask you then check if it works as I don't have a prana with sensors?

**Reply by corvis** [reply]
Also, could you please clarify exact model you have. I'm not sure I see Prana 160 among available models (at least on the Ukrainian market), [there are 150 and 200](https://prana.org.ua/models). Which differ just by power and size, however there each are available in 2 variations - Eco Energy and Eco Life which have a bunch of sensors. I suspect you should have one of them.

According to the description apart from the temperature and humidity sensors there must be also VOC and Co2 sensors. Do you know if your model have them?

**Reply by francesco-re-1107** [reply]
Yeah, I've tested it for a few days and it worked without any issue.

I have [this model](https://www.prana24.com/recuperatore-prana-160-erp/ ) (maybe it's just sold in Italy?) which doesn't have co2 and VOC sensors.


Probably before merging my changes it's a good idea to add some checks in the code cause I don't know if the returned Prana state array length is the same across all devices.

**Reply by corvis** [reply]
Yep this sounds about right... Prana 160 is not available for Ukrainian market(most likely due to different standards for ventilation equipment).

As for the testing, sorry, I meant could you please check the build I'll create after incorporating your changes. As I'll definitely verify compatibility with devices without sensors but won't be able to check if the code still works for ones like yours.

**Reply by francesco-re-1107** [reply]
Ah yeah, I'll test it for sure!

**Reply by corvis** [reply]
Hey @francesco-re-1107,

I added sensors support to the mainline code base. The changes are available in release 0.4.0 ([diff is here](https://github.com/corvis/prana_rc/compare/b2514c94b9b15623567d46da087e040378746ffa..04a1c3d8f1185d310c7863abda7995a57d376d37)). PyPi package is already available so you could check easlily.

The notable difference comparing to your code is the json structure - I wrapped all sensors related properties under `sensors`. See [corresponding documentation section](https://github.com/corvis/prana_rc#get-state)

I could confirm the new release is fully compatible with baseline Prana series (no sensors) but it would be awesome if you verify it works with your Prana 160. Also if possible please try to run it with `PRANA_LOGGING=true` env variable and post output starting from `DEBUG: State data:` here so could create unit tests for devices I don't have in future.

Example:

Run it like `PRANA_LOGGING=true prana -v -d XX:XX:XX:XX:XX:XX status`
And watch for something like:

```
2021-03-04 20:10:22,330 PranaDevice DEBUG: State data:
2021-03-04 20:10:22,330 PranaDevice DEBUG: 0xbe,0xef,0x05,0x01,0x00,0x00,0x00,0x00,0x5a,0xc0,0x01,0xc0,0x20,0xc0,0x01,0xc0,0x00,0xc0,0x00,0x00,0x00,0xc0,0x01,0x80,0x01,0xc0,0x14,0x00,0x01,0x40,0x14,0x00,0x01,0x40,0x14,0x00,0x00,0x80,0x00,0x00,0x00,0xc0,0x00,0x00,0x01,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0xc0,0x00,0x00,0x00,0x00,0x00,0x8f,0xd3,0x3e,0xde,0x67,0xf4,0xfe,0x3a,0xfc,0x34,0x6e,0xde,0xca,0x66,0xa4,0xca,0x73,0xed,0xc5,0x0f,0xeb,0xca,0xd3,0xc8,0x29,0x2e,0x1d,0x25,0x1a,0xce,0x01,0x64,0x65,0x16,0x7c,0xdd,0x23,0x73,0xcc,0xde,0x7c,0x3a,0x06,0xc7,0x6c,0x1d,0x17,0xe2,0x6c,0x34,0x76,0x31,0x1a,0x88,0x38,0xc7,0x67,0x19,0x4c,0xc6
```

**Reply by francesco-re-1107** [reply]
Hi @corvis,
I've just tested it out and it works great.

Here's the log

`PranaDevice DEBUG: 0xbe,0xef,0x05,0x01,0x00,0x00,0x00,0x00,0x5a,0xc0,0x01,0xc0,0x01,0xc0,0x00,0xc0,0x01,0xc0,0x00,0x00,0x01,0xc0,0x00,0x40,0x01,0x40,0x0a,0xc0,0x01,0xc0,0x0a,0xc0,0x01,0xc0,0x0a,0x00,0x00,0x80,0x00,0x00,0x00,0xc0,0x00,0x00,0x01,0x00,0x00,0x00,0x80,0xc9,0x00,0x00,0x00,0x00,0x80,0xbd,0x00,0x00,0x00,0x00,0xbf,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x82,0xdd,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x80,0x01,0x00,0x00,0x00,0x00,0x00,0x00,0xc0,0x00,0x00,0x00,0x00,0x00,0x52,0x25,0xd3,0xe5,0x5f,0xe4,0x97,0x36,0x39,0x93,0xc6,0x25,0xa2,0xcd,0xd1,0x2a,0x28,0x18,0x35,0x78,0x6c,0xd2,0xfc,0x34,0xda,0x11,0x5b,0xc7,0xdd,0x92,0x1d,0xe1,0x26,0x2d,0xfc,0x43,0x39,0x09,0x11,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0xf4,0x1d,0x89,0x70,0xd1,0xb9,0xed,0x3f,0x5a,0x9e,0xd1,0x58,0x02,0xd6`

Great job :)

**Comment by alexliit** [comment]
Hello
After setting up sensor readings on my two prana 160 ERP PRO Mini, now I'm trying to control them.
I was successful to set it with the "prana.set_state" command by this are my consideration / observation

1. speed "off" property is not accepted (webservice call throw a parse error): 0 (zero) works instead
2. speed is not type "string" but interger (webservice call throw a parse error)
3. is it possibile to set manual / auto mode ?
4. is it possibile to se "in" fan at different speed than "out" fan? Speed is alway putting the device in "lock" and "manual" fan speed mode (fan in and fan out at same speed)

Thanks!

**Comment by adysxp** [comment]
Good evening, on new models, prana.get_state/prana.set_state doesn't work(Prana 150 premium Plus, the one with 6 sensors and wifi mode)
